lib/auth/appauthconfig/appauthconfigv1/sessions_service.go

```diff
@@ -0,0 +1,271 @@
+		return nil, trace.Wrap(err)
+	}
+
+	return &appauthconfigv1.CreateAppSessionWithJwtResponse{Session: ws.(*types.WebSessionV2)}, nil
+}
+
+// retrieveJWKSAppAuthConfig retrieves JWKS contents given a app auth JWT
+// config.
+func retrieveJWKSAppAuthConfig(jwtConfig *appauthconfigv1.AppAuthConfigJWTSpec, httpClient *http.Client) (*jose.JSONWebKeySet, []jose.SignatureAlgorithm, error) {
+	var rawJwks []byte
+	switch jwksSource := jwtConfig.KeysSource.(type) {
+	case *appauthconfigv1.AppAuthConfigJWTSpec_JwksUrl:
+		req, err := http.NewRequest(http.MethodGet, jwksSource.JwksUrl, nil)
+		if err != nil {
+			return nil, nil, trace.Wrap(err)
+		}
+
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			return nil, nil, trace.Wrap(err)
+		}
+		defer resp.Body.Close()
+
+		rawJwks, err = utils.ReadAtMost(resp.Body, teleport.MaxHTTPRequestSize)
+		if err != nil {
+			return nil, nil, trace.Wrap(err)
+		}
+	case *appauthconfigv1.AppAuthConfigJWTSpec_StaticJwks:
+		rawJwks = []byte(jwksSource.StaticJwks)
+	}
+
+	var jwks jose.JSONWebKeySet
+	if err := json.Unmarshal(rawJwks, &jwks); err != nil {
+		return nil, nil, trace.Wrap(err, "unable to parse JWKS contents")
+	}
+
+	sigs := make([]jose.SignatureAlgorithm, len(jwks.Keys))
+	for i, jwk := range jwks.Keys {
+		sigs[i] = jose.SignatureAlgorithm(jwk.Algorithm)
+	}
+
+	if len(sigs) == 0 {
+		return nil, nil, trace.BadParameter("empty JWKS contents")
+	}
+
+	return &jwks, sigs, nil
+}
+
+// verifyAppAuthJWTToken verifies the provided JWT token using app auth config
+// and returns the extracted username from JWT claims.
+func verifyAppAuthJWTToken(jwtToken string, jwks *jose.JSONWebKeySet, sigs []jose.SignatureAlgorithm, jwtConfig *appauthconfigv1.AppAuthConfigJWTSpec) (string, time.Time, error) {
+	parsedJwt, err := jwt.ParseSigned(jwtToken, sigs)
+	if err != nil {
+		return "", time.Time{}, trace.Wrap(err)
+	}
+
+	var (
+		claims          jwt.Claims
+		remainingClaims map[string]any
+	)
+	if err := parsedJwt.Claims(jwks, &claims, &remainingClaims); err != nil {
+		return "", time.Time{}, trace.Wrap(err)
+	}
+
+	expected := jwt.Expected{
+		Issuer:      jwtConfig.Issuer,
+		AnyAudience: jwt.Audience{jwtConfig.Audience},
+		Time:        time.Now(),
+	}
+	if err := claims.Validate(expected); err != nil {
+		return "", time.Time{}, trace.Wrap(err)
+	}
+
+	if claims.IssuedAt == nil {
+		return "", time.Time{}, trace.BadParameter("token must have 'iat' claim")
+	}
+
+	if claims.Expiry == nil {
+		return "", time.Time{}, trace.BadParameter("token must have 'exp' claim")
+	}
+
+	issuedAt := claims.IssuedAt.Time()
+	if time.Since(issuedAt) > jwtMaxIssuedAtAfter {
+		return "", time.Time{}, trace.BadParameter("token must be issued recently to be used")
+	}
+
+	usernameClaimName := cmp.Or(jwtConfig.UsernameClaim, defaultUsernameClaim)
+	usernameClaim, ok := remainingClaims[usernameClaimName]
+	if !ok {
+		return "", time.Time{}, trace.BadParameter("token must have %q claim", usernameClaim)
```

Comment:
```suggestion
		return "", time.Time{}, trace.BadParameter("token must have %q claim", usernameClaimName)
```

Shouldn't it be `usernameClaimName` since that is what we're looking up with?